### PR TITLE
[Sigmas] Keep sigmas on CPU

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

After a second round of testing, it seems like keeping the sigmas always on CPU is actually not a bad idea because there is no real slow-down for the "default" mode while the speed-up for torch.compile is significantly higher when leaving sigmas on CPU.
cc @sayakpaul

### CODE REVIEW

File: src/diffusers/schedulers/scheduling_consistency_decoder.py [hunk 1]
```diff
@@ -26,7 +26,8 @@ def betas_for_alpha_bar(
 
     Args:
         num_diffusion_timesteps (`int`): the number of betas to produce.
-        max_beta (`float`): the maximum beta to use; use values lower than 1 to
+        max_beta (`float`): the maximum beta to use
+        use values lower than 1 to
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_consistency_decoder.py:30]:
```suggestion
        max_beta (`float`): the maximum beta to use; use values lower than 1 to
```

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 2]
```diff
@@ -63,7 +63,8 @@ def betas_for_alpha_bar(
 
     Args:
         num_diffusion_timesteps (`int`): the number of betas to produce.
-        max_beta (`float`): the maximum beta to use; use values lower than 1 to
+        max_beta (`float`): the maximum beta to use
+        use values lower than 1 to
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddim.py:67]:
```suggestion
        max_beta (`float`): the maximum beta to use; use values lower than 1 to
```

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 3]
```diff
@@ -161,7 +162,8 @@ class DDIMScheduler(SchedulerMixin, ConfigMixin):
             `set_alpha_to_one=False` to make the last step use step 0 for the previous alpha product like in Stable
             Diffusion.
         prediction_type (`str`, defaults to `epsilon`, *optional*):
-            Prediction type of the scheduler function; can be `epsilon` (predicts the noise of the diffusion process),
+            Prediction type of the scheduler function
+            can be `epsilon` (predicts the noise of the diffusion process),
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddim.py:166]:
```suggestion
            Prediction type of the scheduler function. Can be `epsilon` (predicts the noise of the diffusion process),
```

File: src/diffusers/schedulers/scheduling_ddim_inverse.py [hunk 4]
```diff
@@ -61,7 +61,8 @@ def betas_for_alpha_bar(
 
     Args:
         num_diffusion_timesteps (`int`): the number of betas to produce.
-        max_beta (`float`): the maximum beta to use; use values lower than 1 to
+        max_beta (`float`): the maximum beta to use
+        use values lower than 1 to
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddim_inverse.py:65]:
```suggestion
        max_beta (`float`): the maximum beta to use. Use values lower than 1 to
```

File: src/diffusers/schedulers/scheduling_ddim_inverse.py [hunk 5]
```diff
@@ -159,7 +160,8 @@ class DDIMInverseScheduler(SchedulerMixin, ConfigMixin):
             `set_alpha_to_one=False` to make the last step use `num_train_timesteps - 1` for the previous alpha
             product.
         prediction_type (`str`, defaults to `epsilon`, *optional*):
-            Prediction type of the scheduler function; can be `epsilon` (predicts the noise of the diffusion process),
+            Prediction type of the scheduler function
+            can be `epsilon` (predicts the noise of the diffusion process),
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddim_inverse.py:164]:
```suggestion
            Prediction type of the scheduler function. Can be `epsilon` (predicts the noise of the diffusion process),
```

File: src/diffusers/schedulers/scheduling_ddim_parallel.py [hunk 6]
```diff
@@ -63,7 +63,8 @@ def betas_for_alpha_bar(
 
     Args:
         num_diffusion_timesteps (`int`): the number of betas to produce.
-        max_beta (`float`): the maximum beta to use; use values lower than 1 to
+        max_beta (`float`): the maximum beta to use
+        use values lower than 1 to
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddim_parallel.py:67]:
```suggestion
        max_beta (`float`): the maximum beta to use. Use values lower than 1 to
```

File: src/diffusers/schedulers/scheduling_ddpm.py [hunk 7]
```diff
@@ -60,7 +60,8 @@ def betas_for_alpha_bar(
 
     Args:
         num_diffusion_timesteps (`int`): the number of betas to produce.
-        max_beta (`float`): the maximum beta to use; use values lower than 1 to
+        max_beta (`float`): the maximum beta to use
+        use values lower than 1 to
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddpm.py:64]:
```suggestion
        max_beta (`float`): the maximum beta to use; Use values lower than 1 to
```

File: src/diffusers/schedulers/scheduling_ddpm.py [hunk 8]
```diff
@@ -114,7 +115,8 @@ class DDPMScheduler(SchedulerMixin, ConfigMixin):
         clip_sample_range (`float`, defaults to 1.0):
             The maximum magnitude for sample clipping. Valid only when `clip_sample=True`.
         prediction_type (`str`, defaults to `epsilon`, *optional*):
-            Prediction type of the scheduler function; can be `epsilon` (predicts the noise of the diffusion process),
+            Prediction type of the scheduler function
+            can be `epsilon` (predicts the noise of the diffusion process),
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddpm.py:119]:
```suggestion
            Prediction type of the scheduler function. Can be `epsilon` (predicts the noise of the diffusion process),
```

File: src/diffusers/schedulers/scheduling_ddpm_parallel.py [hunk 9]
```diff
@@ -62,7 +62,8 @@ def betas_for_alpha_bar(
 
     Args:
         num_diffusion_timesteps (`int`): the number of betas to produce.
-        max_beta (`float`): the maximum beta to use; use values lower than 1 to
+        max_beta (`float`): the maximum beta to use
+        use values lower than 1 to
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddpm_parallel.py:66]:
```suggestion
        max_beta (`float`): the maximum beta to use. Use values lower than 1 to
```

File: src/diffusers/schedulers/scheduling_ddpm_wuerstchen.py [hunk 10]
```diff
@@ -56,7 +56,8 @@ def betas_for_alpha_bar(
 
     Args:
         num_diffusion_timesteps (`int`): the number of betas to produce.
-        max_beta (`float`): the maximum beta to use; use values lower than 1 to
+        max_beta (`float`): the maximum beta to use
+        use values lower than 1 to
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddpm_wuerstchen.py:60]:
```suggestion
        max_beta (`float`): the maximum beta to use. Use values lower than 1 to
```

File: src/diffusers/schedulers/scheduling_deis_multistep.py [hunk 11]
```diff
@@ -42,7 +42,8 @@ def betas_for_alpha_bar(
 
     Args:
         num_diffusion_timesteps (`int`): the number of betas to produce.
-        max_beta (`float`): the maximum beta to use; use values lower than 1 to
+        max_beta (`float`): the maximum beta to use
+        use values lower than 1 to
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_deis_multistep.py:46]:
```suggestion
        max_beta (`float`): the maximum beta to use. Use values lower than 1 to
```

File: src/diffusers/schedulers/scheduling_deis_multistep.py [hunk 12]
```diff
@@ -94,7 +95,8 @@ class DEISMultistepScheduler(SchedulerMixin, ConfigMixin):
             The DEIS order which can be `1` or `2` or `3`. It is recommended to use `solver_order=2` for guided
             sampling, and `solver_order=3` for unconditional sampling.
         prediction_type (`str`, defaults to `epsilon`):
-            Prediction type of the scheduler function; can be `epsilon` (predicts the noise of the diffusion process),
+            Prediction type of the scheduler function
+            can be `epsilon` (predicts the noise of the diffusion process),
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_deis_multistep.py:99]:
```suggestion
            Prediction type of the scheduler function. Can be `epsilon` (predicts the noise of the diffusion process),
```

File: src/diffusers/schedulers/scheduling_lcm.py [hunk 13]
```diff
@@ -250,6 +250,7 @@ def __init__(
         self.custom_timesteps = False
 
         self._step_index = None
+        self.sigmas.to('cpu') # to avoid too much CPU/GPU communication
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_lcm.py:253]:
```suggestion
```

File: src/diffusers/schedulers/scheduling_lcm.py [hunk 14]
```diff
@@ -462,6 +463,7 @@ def set_timesteps(
         self.timesteps = torch.from_numpy(timesteps).to(device=device, dtype=torch.long)
 
         self._step_index = None
+        self.sigmas.to('cpu') # to avoid too much CPU/GPU communication
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_lcm.py:466]:
```suggestion
```

File: src/diffusers/schedulers/scheduling_ipndm.py [hunk 15]
```diff
@@ -56,6 +56,7 @@ def __init__(
         # running values
         self.ets = []
         self._step_index = None
+        self.sigmas.to('cpu') # to avoid too much CPU/GPU communication
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ipndm.py:59]:
```suggestion
```

File: src/diffusers/schedulers/scheduling_ipndm.py [hunk 16]
```diff
@@ -90,6 +91,7 @@ def set_timesteps(self, num_inference_steps: int, device: Union[str, torch.devic
 
         self.ets = []
         self._step_index = None
+        self.sigmas.to('cpu') # to avoid too much CPU/GPU communication
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ipndm.py:94]:
```suggestion
```

**Review by yiyixuxu — APPROVED** [review]
thanks!

**Review by sayakpaul — APPROVED** [review]
Thank you!

**Comment by sunhs** [comment]
I'm curious about how it works. `self.sigmas.to("cpu")` does return a Tensor on cpu but it's not used, whereas `self. sigmas` is still on gpu. I must have misunderstood something. Could you show me some references to make it clear?

**Comment by sayakpaul** [comment]
to is inplace.

**Comment by sunhs** [comment]
image

Can this demonstrate my uncertainty?

**Comment by sayakpaul** [comment]
Cc: @patrickvonplaten here. This still doesn't affect the results in the PyTorch blog post as it was done from a separate branch with another way of doing the device placement.

**Comment by yiyixuxu** [comment]
ahh does this means our `pipeline.to()` behave differently from the pytorch `to()`? - ours would modify in place and return a copy

**Comment by sunhs** [comment]
> ahh does this means our `pipeline.to()` behave differently from the pytorch `to()`? - ours would modify in place and return a copy

This should not be a problem, since `nn.Module.to` indeed is in place while `Tensor.to` isn't.

**Comment by patrickvonplaten** [comment]
Nice catch @sunhs - correcting it here: https://github.com/huggingface/diffusers/pull/6708 

It shouldn't change anything in practice since `self.sigmas` are always kept on CPU anyways for Euler and DPM. I'll keep an eye on the slow tests though (cc @DN6)

**Pull request merged**